### PR TITLE
[libshortfin] Add device_array.view

### PR DESCRIPTION
**Pull request by Groverkss**
Currently only restricted to row-major dense array views.

### CODE REVIEW

File: libshortfin/python/array_binding.cc [hunk 1]
```diff
@@ -30,6 +30,14 @@ static const char DOCSTRING_ARRAY_FILL[] = R"(Fill an array with a value.
 Equivalent to `array.storage.fill(pattern)`.
 )";
 
+static const char DOCSTRING_ARRAY_VIEW[] =
+    R"(Create a view of an array.
+
+Either integer indices or slices can be passed to the view() method to create
```
Comment (stellaraccident) [inline libshortfin/python/array_binding.cc:36]: Afaict, this statement isn't correct since it is taking an explicit offset/sizes args vs a list that can either be int or slice. The latter is more compatible with prevailing Python norms, fwiw.

File: libshortfin/tests/api/array_test.py [hunk 2]
```diff
@@ -121,3 +121,13 @@ def test_xtensor_types(scope, dtype, code, py_value, expected_str):
     r = repr(ary)
     print("__repr__ =", r)
     assert expected_str in r, f"Expected '{expected_str}' in '{r}'"
+
+def test_view(lsys, device):
+    async def main():
+        src = sfnp.device_array(device, [2, 4], sfnp.uint8)
+        src.fill(b"\0\1\2\3")
+
+        view1 = src.view(offsets=[0, 1], sizes=[1, 3])
```
Comment (stellaraccident) [inline libshortfin/tests/api/array_test.py:130]: Some more checking or boundary cases would be good.

File: libshortfin/python/array_binding.cc (hunk 1 above)
Comment (Groverkss) [reply libshortfin/python/array_binding.cc:36]: int/slice makes sense. I switched back to that.

File: libshortfin/tests/api/array_test.py (hunk 2 above)
Comment (Groverkss) [reply libshortfin/tests/api/array_test.py]: Added some more tests

**Comment by Groverkss** [comment]
The CI failure seems to be weird. It seems to be pointing at the array not being filled properly. @marbre Do you know if this is a CI flake?

**Comment by stellaraccident** [comment]
There's nothing funny with scheduling there or anything. All synchronous code. 

I'd throw an assert after the fill to verify that is working. It's all pretty new code.

File: libshortfin/tests/api/array_test.py [hunk 3]
```diff
@@ -121,3 +121,34 @@ def test_xtensor_types(scope, dtype, code, py_value, expected_str):
...
+        # Row indexing
+        ([0], "{{0, 1, 2, 3}}"),
+        # Sliced indexing
+        ([1, slice(2, 4)], "{{2, 3}}"),
+        ([slice(1, 2), slice(2, 4)], "{{2, 3}}"),
+    ],
+)
+def test_view(scope, keys, expected_str):
+    src = sfnp.device_array.for_host(scope.device(0), [4, 4], sfnp.uint8)
+    src.fill(b"\0\1\2\3")
```
Comment (stellaraccident) [inline libshortfin/tests/api/array_test.py:140]: Actually, I misspoke: `fill` is asynchronous, so you are observing the value before the queue has processed the commands. This needs some more ergonomic work: we generally don't want to be async filling a host buffer like this for this kind of case. And I'm missing synchronous waits in the Python API, which would be the most convenient here.

Easiest thing would be if the items= API was available.

Best thing to do for now is to write this async like the tests above. Look at `test_fill_copy_to_for_transfer`. Note the `await device` before observing the result.

There will be better ergonomics for this eventually but no need to wait for them for a test.

**Review by stellaraccident — APPROVED** [review]
I left a note on how to fix the test. If that works, lgtm and can land.

File: libshortfin/tests/api/array_test.py (hunk 3 above)
Comment (Groverkss) [reply libshortfin/tests/api/array_test.py]: ah, added an await to fix it.
